src/unix/fs.c

```diff
@@ -766,6 +770,102 @@ static ssize_t uv__fs_write(uv_fs_t* req) {
   return r;
 }
 
+static ssize_t uv__fs_copyfile(uv_fs_t* req) {
+#if defined(__APPLE__) && !TARGET_OS_IPHONE
+  /* On macOS, use the native copyfile(3). */
+  copyfile_flags_t flags;
+
+  flags = COPYFILE_ALL;
+
+  if (req->flags & UV_FS_COPYFILE_EXCL)
+    flags = flags | COPYFILE_EXCL;
+
+  return copyfile(req->path, req->new_path, NULL, flags);
+#else
+  uv_fs_t fs_req;
+  uv_file src;
+  uv_file dst;
+  struct stat statsbuf;
+  int dst_flags;
+  int result;
+  int err;
+
+  dst = -1;
+
+  /* Open the source file. */
+  src = uv_fs_open(req->loop, &fs_req, req->path, O_RDONLY, 0, NULL);
+  uv_fs_req_cleanup(&fs_req);
+
+  if (src < 0) {
+    err = src;
+    goto out;
+  }
+
+  /* Get the source file's mode. */
+  err = fstat(src, &statsbuf);
+
+  if (err != 0) {
+    err = -errno;
+    goto out;
+  }
+
+  dst_flags = O_WRONLY | O_CREAT;
+
+  if (req->flags & UV_FS_COPYFILE_EXCL)
+    dst_flags = dst_flags | O_EXCL;
+
+  /* Open the destination file. */
+  dst = uv_fs_open(req->loop,
+                   &fs_req,
+                   req->new_path,
+                   dst_flags,
+                   statsbuf.st_mode,
+                   NULL);
+  uv_fs_req_cleanup(&fs_req);
+
+  if (dst < 0) {
+    err = dst;
+    goto out;
+  }
+
+  err = uv_fs_sendfile(req->loop, &fs_req, dst, src, 0, statsbuf.st_size, NULL);
+  uv_fs_req_cleanup(&fs_req);
+out:
+  if (err < 0)
+    result = err;
+  else
+    result = 0;
+
+  /* Close the source and destination files. */
+  if (src >= 0) {
+    err = uv_fs_close(NULL, &fs_req, src, NULL);
+    uv_fs_req_cleanup(&fs_req);
+
+    /* Don't overwrite any existing errors. */
+    if (err != 0 && result == 0)
+      result = err;
+  }
+
+  if (dst >= 0) {
+    err = uv_fs_close(NULL, &fs_req, dst, NULL);
+    uv_fs_req_cleanup(&fs_req);
```

Comment:
@bnoordhuis just to clarify. You mean use `uv__close_nocheckstdio()` instead of `uv_fs_close()` and the cleanup? And if so, by "here and below" you mean "here and above"?